ignite/cmd/chain_serve.go

```diff
@@ -72,82 +80,123 @@ production, you may want to run "appd start" manually.
+	} else {
+		options = append(options, cliui.WithVerbosity(verbosity))
 	}
 
-	if resetOnce {
-		serveOptions = append(serveOptions, chain.ServeResetOnce())
-	}
+	session := cliui.New(options...)
+	defer session.End()
 
-	quitOnFail, err := cmd.Flags().GetBool(flagQuitOnFail)
-	if err != nil {
-		return err
-	}
+	// Depending on the verbosity execute the serve command within
+	// a bubbletea context to display the custom UI.
+	serve := chainServeCmd(cmd, session)
+	if verbosity == uilog.VerbosityDefault {
+		bus := session.EventBus()
+		bus.Send("Initializing...", events.ProgressStart())
 
-	if quitOnFail {
-		serveOptions = append(serveOptions, chain.QuitOnFail())
+		// Render UI
+		m := cmdmodel.NewChainServe(cmd, bus, serve)
+		if err := tea.NewProgram(m).Start(); err != nil {
+			return err
+		}
 	}
 
-	generateClients, err := cmd.Flags().GetBool(flagGenerateClients)
-	if err != nil {
-		return err
+	// Otherwise run the serve command directly
+	if msg, ok := serve().(cliuimodel.ErrorMsg); ok {
+		return msg.Error
 	}
 
-	if generateClients {
-		serveOptions = append(serveOptions, chain.GenerateClients())
-	}
+	return nil
+}
 
-	if flagGetSkipProto(cmd) {
-		serveOptions = append(serveOptions, chain.ServeSkipProto())
+func chainServeCmd(cmd *cobra.Command, session *cliui.Session) tea.Cmd {
+	return func() tea.Msg {
+		chainOption := []chain.Option{
+			chain.WithOutputer(session),
+			chain.CollectEvents(session.EventBus()),
+		}
+
+		if flagGetCheckDependencies(cmd) {
+			chainOption = append(chainOption, chain.CheckDependencies())
+		}
+
+		// check if custom config is defined
+		config, err := cmd.Flags().GetString(flagConfig)
+		if err != nil {
+			return err
+		}
+		if config != "" {
+			chainOption = append(chainOption, chain.ConfigFile(config))
+		}
+
+		// create the chain
+		c, err := NewChainWithHomeFlags(cmd, chainOption...)
+		if err != nil {
+			return err
+		}
+
+		cacheStorage, err := newCache(cmd)
+		if err != nil {
+			return err
+		}
+
+		// serve the chain
+		var serveOptions []chain.ServeOption
+
+		forceUpdate, err := cmd.Flags().GetBool(flagForceReset)
+		if err != nil {
+			return err
+		}
+
+		if forceUpdate {
+			serveOptions = append(serveOptions, chain.ServeForceReset())
+		}
+
+		resetOnce, err := cmd.Flags().GetBool(flagResetOnce)
+		if err != nil {
+			return err
+		}
+
+		if resetOnce {
+			serveOptions = append(serveOptions, chain.ServeResetOnce())
+		}
+
+		quitOnFail, err := cmd.Flags().GetBool(flagQuitOnFail)
+		if err != nil {
+			return err
+		}
+
+		if quitOnFail {
+			serveOptions = append(serveOptions, chain.QuitOnFail())
+		}
+
+		generateClients, err := cmd.Flags().GetBool(flagGenerateClients)
+		if err != nil {
+			return err
+		}
+
+		if generateClients {
+			serveOptions = append(serveOptions, chain.GenerateClients())
+		}
+
+		if flagGetSkipProto(cmd) {
+			serveOptions = append(serveOptions, chain.ServeSkipProto())
+		}
+
+		err = c.Serve(cmd.Context(), cacheStorage, serveOptions...)
```

Comment:
If you move that part into a separate function, you can avoid the dirty part where you call `serve().(cliuimodel.ErrorMsg)` when the command is run in verbose mode.

Something like that :
```go
	// Depending on the verbosity execute the serve command within
	// a bubbletea context to display the custom UI.
	if verbosity == uilog.VerbosityDefault {
		bus := session.EventBus()
		bus.Send("Initializing...", events.ProgressStart())

		// Render UI
		serve := chainServeCmd(cmd, session)
		m := cmdmodel.NewChainServe(cmd, bus, serve)
		if err := tea.NewProgram(m).Start(); err != nil {
			return err
		}
	}

	// Otherwise run the serve command directly
	return chainServe(cmd, session)
}

func chainServeCmd(cmd *cobra.Command, session *cliui.Session) tea.Cmd {
	return func() tea.Msg {
		if err := chainServe(cmd, session); err != nil {
			return cliuimodel.ErrorMsg{Error: err}
		}
		return cliuimodel.QuitMsg{}
	}
}

func chainServe(cmd *cobra.Command, session *cliui.Session) error {
	(...)
}
```

Reply:
Changeset d3469d8